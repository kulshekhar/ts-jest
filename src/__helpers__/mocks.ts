import { testing } from 'bs-logger'

import { rootLogger } from '../util/logger'

export const logTargetMock = () => (rootLogger as testing.LoggerMock).target

export const mockObject = <T, M>(obj: T, newProps: M): T & M & { mockRestore: () => T } => {
  const backup: any = Object.create(null)

  Object.keys(newProps).forEach(key => {
    const desc = (backup[key] = Object.getOwnPropertyDescriptor(obj, key))
    const newDesc: any = { ...desc }
    if (newDesc.get) {
      newDesc.get = () => (newProps as any)[key]
    } else {
      newDesc.value = (newProps as any)[key]
    }
    Object.defineProperty(obj, key, newDesc)
  })
  if ((obj as any).mockRestore) backup.mockRestore = Object.getOwnPropertyDescriptor(obj, 'mockRestore')
  return Object.defineProperty(obj, 'mockRestore', {
    value() {
      Object.keys(backup).forEach(key => {
        Object.defineProperty(obj, key, backup[key])
      })
      return obj
    },
    configurable: true,
  })
}

export const mockWriteStream = () => ({
  written: [] as string[],
  write(text: string) {
    this.written.push(text)
  },
  clear() {
    this.written = []
  },
})
