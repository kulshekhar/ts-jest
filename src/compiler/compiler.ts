import * as ts from 'typescript';
import * as fs from 'fs';
import * as path from 'path';
import * as utils from './utils';

function formatTscParserErrors(errors: ts.Diagnostic[]) {
  return errors.map(s => JSON.stringify(s, null, 4)).join('\n');
}

interface File {
  path: string;
  version: number;
  snapshot: ts.IScriptSnapshot;
  src: string;
}

class Files {
  private files: Map<string, File>;
  constructor() {
    this.files = new Map();
  }
  // set(path: string, file: File) {
  //     this.files.set(path, file)
  // }
  update({ path, src }: { path: string; src: string }) {
    const file = this.get(path);
    if (file === undefined) {
      const newFile: File = {
        src,
        path,
        version: 0,
        snapshot: ts.ScriptSnapshot.fromString(src),
      };
      this.files.set(path, newFile);
      return;
    }
    if (file.src === src) {
      return;
    }
    file.src = src;
    file.version += 1;
    file.snapshot = ts.ScriptSnapshot.fromString(src);
  }
  get(path: string) {
    return this.files.get(path);
  }
  getFileNames() {
    return Array.from(this.files.values()).map(f => f.path);
  }
  getScriptVersion(path: string) {
    const f = this.get(path);
    if (f === undefined) {
      return '';
    }
    return f.version.toString();
  }
}

export class Compiler {
  private options?: ts.CompilerOptions;
  private service: ts.LanguageService;
  private files: Files;

  constructor(options?: ts.CompilerOptions) {
    this.files = new Files();
    this.options = options;
    this.service = this.createServiceHost();
  }
  setOptions(options: ts.CompilerOptions) {
    this.options = options;
  }
  private createServiceHost() {
    const that = this;

    let service: ts.LanguageService;

    class ServiceHost implements ts.LanguageServiceHost {
      // getCustomTransformers() {
      //     return {
      //         before: [tsruntimeTransformer(service.getProgram())]
      //     }
      // }
      getScriptFileNames() {
        return that.files.getFileNames();
      }
      getScriptVersion(fileName: string) {
        return that.files.getScriptVersion(fileName);
      }
      getScriptSnapshot(fileName: string) {
        const file = that.files.get(fileName);
        if (file !== undefined) {
          return file.snapshot;
        }
        if (!fs.existsSync(fileName)) {
          return undefined;
        }
        return ts.ScriptSnapshot.fromString(
          fs.readFileSync(fileName).toString(),
        ); //todo maybe put to files cache
      }
      getCurrentDirectory = () => process.cwd();
      getCompilationSettings = () => that.options;
      getDefaultLibFileName = (options: ts.CompilerOptions) =>
        ts.getDefaultLibFilePath(options);
      fileExists = ts.sys.fileExists;
      readFile = ts.sys.readFile;
      readDirectory = ts.sys.readDirectory;
      // resolveTypeReferenceDirectives(typeDirectiveNames: string[], containingFile: string) {
      //     const resolved = typeDirectiveNames.map(directive =>
      //         ts.resolveTypeReferenceDirective(directive, containingFile, options, ts.sys)
      //             .resolvedTypeReferenceDirective);

      //     // resolved.forEach(res => {
      //     if (res && res.resolvedFileName) {
      //         fileDeps.add(containingFile, res.resolvedFileName);
      //     }
      // });

      // return resolved;
      // }
    }

    service = ts.createLanguageService(
      new ServiceHost(),
      ts.createDocumentRegistry(),
    );

    return service;
  }

  emitFile({ path, src }: { path: string; src: string }) {
    this.files.update({ path, src });

    let output = this.service.getEmitOutput(path);

    if (output.emitSkipped) {
      const errors = this.formatErrors(path);
      throw new Error(`Errors while compile ts: \n${errors}`);
    }
    const res = utils.findResultFor(path, output);
    return res;
  }

  private formatErrors(fileName: string) {
    let allDiagnostics = this.service
      .getCompilerOptionsDiagnostics()
      .concat(this.service.getSyntacticDiagnostics(fileName))
      .concat(this.service.getSemanticDiagnostics(fileName));

    let res = '';
    allDiagnostics.forEach(diagnostic => {
      let message = ts.flattenDiagnosticMessageText(
        diagnostic.messageText,
        '\n',
      );
      if (diagnostic.file) {
        let { line, character } = diagnostic.file.getLineAndCharacterOfPosition(
          diagnostic.start!,
        );
        res += `  Error ${diagnostic.file.fileName} (${line + 1},${character +
          1}): ${message}\n`;
      } else {
        res += `  Error: ${message}\n`;
      }
    });
    return res;
  }
}
