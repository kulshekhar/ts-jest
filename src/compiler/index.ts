export * from './compiler-utils'
export * from './ts-compiler'
export * from './ts-jest-compiler'
