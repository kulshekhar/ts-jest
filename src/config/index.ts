export * from './config-set'
export * from './paths-to-module-name-mapper'
