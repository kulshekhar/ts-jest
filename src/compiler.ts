/**
 * This code is heavilly inspired from
 * https://github.com/JsCommunity/make-error/blob/v1.3.4/index.js
 * ...but more modified than expected :-D
 * Below is the original license anyway:
 *
 * ---
 *
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Blake Embrey (hello@blakeembrey.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import { LogContexts, LogLevels, Logger } from 'bs-logger'
import bufferFrom = require('buffer-from')
import stableStringify = require('fast-json-stable-stringify')
import { readFileSync, writeFileSync } from 'fs'
import memoize = require('lodash.memoize')
import mkdirp = require('mkdirp')
import { basename, extname, join, normalize, relative } from 'path'
import { LanguageServiceHost } from 'typescript'

import { ConfigSet } from './config/config-set'
import { MemoryCache, TsCompiler, TypeInfo } from './types'
import { Errors, interpolate } from './util/messages'
import { sha1 } from './util/sha1'

const hasOwn = Object.prototype.hasOwnProperty

/**
 * Register TypeScript compiler.
 * @internal
 */
export function createCompiler(configs: ConfigSet): TsCompiler {
  const logger = configs.logger.child({ namespace: 'ts-compiler' })
  logger.debug(
    'creating typescript compiler',
    configs.tsJest.isolatedModules ? '(isolated modules)' : '(language service)',
  )
  const cachedir = configs.tsCacheDir

  const memoryCache: MemoryCache = {
    contents: Object.create(null),
    versions: Object.create(null),
    outputs: Object.create(null),
  }

  // Require the TypeScript compiler and configuration.
  const ts = configs.compilerModule
  const cwd = configs.cwd

  const extensions = ['.ts', '.tsx']
  const {
    typescript: { options: compilerOptions, fileNames },
  } = configs

  // Enable `allowJs` when flag is set.
  if (compilerOptions.allowJs) {
    extensions.push('.js')
    extensions.push('.jsx')
  }

  // Initialize files from TypeScript into project.
  for (const path of fileNames) memoryCache.versions[path] = 1

  /**
   * Get the extension for a transpiled file.
   */
  const getExtension =
    compilerOptions.jsx === ts.JsxEmit.Preserve
      ? (path: string) => (/\.[tj]sx$/.test(path) ? '.jsx' : '.js')
      : (_: string) => '.js'

  const transformers = configs.tsCustomTransformers

  /**
   * Create the basic required function using transpile mode.
   */
  let getOutput = (code: string, fileName: string /* , lineOffset = 0 */): SourceOutput => {
    logger.debug({ fileName }, 'getOutput(): compiling as isolated module')
    const result = ts.transpileModule(code, {
      fileName,
      transformers,
      compilerOptions,
      reportDiagnostics: configs.shouldReportDiagnostic(fileName),
    })

    if (result.diagnostics) configs.raiseDiagnostics(result.diagnostics, fileName, logger)

    return [result.outputText, result.sourceMapText as string]
  }

  let getTypeInfo = (_code: string, _fileName: string, _position: number): TypeInfo => {
    throw new TypeError(Errors.TypesUnavailableWithoutTypeCheck)
  }

  // Use full language services when the fast option is disabled.
  if (!configs.tsJest.isolatedModules) {
    // Set the file contents into cache.
    const updateMemoryCache = (code: string, fileName: string) => {
      logger.debug({ fileName }, `updateMemoryCache()`)
      if (memoryCache.contents[fileName] !== code) {
        memoryCache.contents[fileName] = code
        memoryCache.versions[fileName] = (memoryCache.versions[fileName] || 0) + 1
      }
    }

    // Create the compiler host for type checking.
    const serviceHostDebugCtx = {
      [LogContexts.logLevel]: LogLevels.debug,
      namespace: 'ts:serviceHost',
      call: null,
    }
    const serviceHostTraceCtx = {
      ...serviceHostDebugCtx,
      [LogContexts.logLevel]: LogLevels.trace,
    }

    const serviceHost: LanguageServiceHost = {
      getScriptFileNames: () => Object.keys(memoryCache.versions),
      getScriptVersion: (fileName: string) => {
        const normalizedFileName = normalize(fileName)
        const version = memoryCache.versions[normalizedFileName]

        // We need to return `undefined` and not a string here because TypeScript will use
        // `getScriptVersion` and compare against their own version - which can be `undefined`.
        // If we don't return `undefined` it results in `undefined === "undefined"` and run
        // `createProgram` again (which is very slow). Using a `string` assertion here to avoid
        // TypeScript errors from the function signature (expects `(x: string) => string`).
        return version === undefined ? ((undefined as any) as string) : String(version)
      },
      getScriptSnapshot(fileName: string) {
        const normalizedFileName = normalize(fileName)
        const hit = hasOwn.call(memoryCache.contents, normalizedFileName)
        logger.trace({ normalizedFileName, cacheHit: hit }, `getScriptSnapshot():`, 'cache', hit ? 'hit' : 'miss')
        // Read contents from TypeScript memory cache.
        if (!hit) {
          memoryCache.contents[normalizedFileName] = ts.sys.readFile(normalizedFileName)
        }

        const contents = memoryCache.contents[normalizedFileName]
        if (contents === undefined) {
          return
        }
        return ts.ScriptSnapshot.fromString(contents)
      },
      fileExists: memoize(ts.sys.fileExists),
      readFile: logger.wrap(serviceHostTraceCtx, 'readFile', memoize(ts.sys.readFile)),
      readDirectory: memoize(ts.sys.readDirectory),
      getDirectories: memoize(ts.sys.getDirectories),
      directoryExists: memoize(ts.sys.directoryExists),
      realpath: memoize(ts.sys.realpath!),
      getNewLine: () => '\n',
      getCurrentDirectory: () => cwd,
      getCompilationSettings: () => compilerOptions,
      getDefaultLibFileName: () => ts.getDefaultLibFilePath(compilerOptions),
      getCustomTransformers: () => transformers,
    }

    logger.debug('creating language service')
    const service = ts.createLanguageService(serviceHost)

    getOutput = (code: string, fileName: string /*, lineOffset = 0 */) => {
      logger.debug({ fileName }, 'getOutput(): compiling using language service')
      // Must set memory cache before attempting to read file.
      updateMemoryCache(code, fileName)

      const output = service.getEmitOutput(fileName)

      if (configs.shouldReportDiagnostic(fileName)) {
        logger.debug({ fileName }, 'getOutput(): computing diagnostics')
        // Get the relevant diagnostics - this is 3x faster than `getPreEmitDiagnostics`.
        const diagnostics = service
          .getCompilerOptionsDiagnostics()
          .concat(service.getSyntacticDiagnostics(fileName))
          .concat(service.getSemanticDiagnostics(fileName))

        // will raise or just warn diagnostics depending on config
        configs.raiseDiagnostics(diagnostics, fileName, logger)
      }

      /* istanbul ignore next (this should never happen but is kept for security) */
      if (output.emitSkipped) {
        throw new TypeError(`${relative(cwd, fileName)}: Emit skipped`)
      }

      // Throw an error when requiring `.d.ts` files.
      /* istanbul ignore next (this should never happen but is kept for security) */
      if (output.outputFiles.length === 0) {
        throw new TypeError(
          interpolate(Errors.UnableToRequireDefinitionFile, {
            file: basename(fileName),
          }),
        )
      }

      return [output.outputFiles[1].text, output.outputFiles[0].text]
    }

    getTypeInfo = (code: string, fileName: string, position: number) => {
      updateMemoryCache(code, fileName)

      const info = service.getQuickInfoAtPosition(fileName, position)
      const name = ts.displayPartsToString(info ? info.displayParts : [])
      const comment = ts.displayPartsToString(info ? info.documentation : [])

      return { name, comment }
    }
  }

  const compile = readThrough(cachedir, memoryCache, getOutput, getExtension, cwd, logger)
  return { cwd, compile, getTypeInfo, extensions, cachedir, ts }
}

/**
 * Internal source output.
 */
type SourceOutput = [string, string]

/**
 * Wrap the function with caching.
 */
function readThrough(
  cachedir: string | undefined,
  memoryCache: MemoryCache,
  compile: (code: string, fileName: string, lineOffset?: number) => SourceOutput,
  getExtension: (fileName: string) => string,
  cwd: string,
  logger: Logger,
) {
  if (!cachedir) {
    return (code: string, fileName: string, lineOffset?: number) => {
      logger.debug({ fileName }, 'readThrough(): no cache')

      const [value, sourceMap] = compile(code, fileName, lineOffset)
      const output = updateOutput(value, fileName, sourceMap, getExtension, cwd)

      memoryCache.outputs[fileName] = output

      return output
    }
  }

  // Make sure the cache directory exists before continuing.
  mkdirp.sync(cachedir)

  return (code: string, fileName: string, lineOffset?: number) => {
    const cachePath = join(cachedir, getCacheName(code, fileName))
    const extension = getExtension(fileName)
    const outputPath = `${cachePath}${extension}`

    try {
      const output = readFileSync(outputPath, 'utf8')
      if (isValidCacheContent(output)) {
        logger.debug({ fileName }, 'readThrough(): cache hit')
        memoryCache.outputs[fileName] = output
        return output
      }
    } catch (err) {}

    logger.debug({ fileName }, 'readThrough(): cache miss')
    const [value, sourceMap] = compile(code, fileName, lineOffset)
    const output = updateOutput(value, fileName, sourceMap, getExtension, cwd)

    logger.debug({ fileName, outputPath }, 'readThrough(): writing caches')
    memoryCache.outputs[fileName] = output
    writeFileSync(outputPath, output)

    return output
  }
}

/**
 * Update the output remapping the source map.
 */
function updateOutput(
  outputText: string,
  fileName: string,
  sourceMap: string,
  getExtension: (fileName: string) => string,
  sourceRoot: string,
) {
  const base = basename(fileName)
  const base64Map = bufferFrom(updateSourceMap(sourceMap, fileName, sourceRoot), 'utf8').toString('base64')
  const sourceMapContent = `data:application/json;charset=utf-8;base64,${base64Map}`
  const sourceMapLength = `${base}.map`.length + (getExtension(fileName).length - extname(fileName).length)

  return outputText.slice(0, -sourceMapLength) + sourceMapContent
}

/**
 * Update the source map contents for improved output.
 */
function updateSourceMap(sourceMapText: string, fileName: string, _sourceRoot: string) {
  const sourceMap = JSON.parse(sourceMapText)
  // const relativeFilePath = posix.normalize(relative(sourceRoot, fileName))
  // sourceMap.file = relativeFilePath
  // sourceMap.sources = [relativeFilePath]
  // sourceMap.sourceRoot = normalize(sourceRoot)
  sourceMap.file = fileName
  sourceMap.sources = [fileName]
  delete sourceMap.sourceRoot
  return stableStringify(sourceMap)
}

/**
 * Get the file name for the cache entry.
 */
function getCacheName(sourceCode: string, fileName: string) {
  return sha1(fileName, '\x00', sourceCode)
}

/**
 * Ensure the given cached content is valid by sniffing for a base64 encoded '}'
 * at the end of the content, which should exist if there is a valid sourceMap present.
 */
function isValidCacheContent(contents: string) {
  return /(?:9|0=|Q==)$/.test(contents.slice(-3))
}
