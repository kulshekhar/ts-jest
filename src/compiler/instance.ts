/**
 * This code is heavily inspired from
 * https://github.com/JsCommunity/make-error/blob/v1.3.4/index.js
 * ...but more modified than expected :-D
 * Below is the original license anyway:
 *
 * ---
 *
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Blake Embrey (hello@blakeembrey.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import { Logger } from 'bs-logger'
import { readFileSync, writeFileSync } from 'fs'
import mkdirp = require('mkdirp')
import { basename, extname, join, normalize } from 'path'

import { ConfigSet } from '../config/config-set'
import { CompileFn, CompilerInstance, MemoryCache, TSFile, TsCompiler } from '../types'
import { sha1 } from '../util/sha1'

import { getResolvedModulesCache } from './compiler-utils'
import { initializeLanguageServiceInstance } from './language-service'
import { initializeProgramInstance } from './program'
import { initializeTranspilerInstance } from './transpiler'

/**
 * Update the output remapping the source map.
 */
function updateOutput(
  outputText: string,
  normalizedFileName: string,
  sourceMap: string,
  getExtension: (fileName: string) => string,
) {
  const base = basename(normalizedFileName)
  const base64Map = Buffer.from(updateSourceMap(sourceMap, normalizedFileName), 'utf8').toString('base64')
  const sourceMapContent = `data:application/json;charset=utf-8;base64,${base64Map}`
  const sourceMapLength =
    `${base}.map`.length + (getExtension(normalizedFileName).length - extname(normalizedFileName).length)

  return outputText.slice(0, -sourceMapLength) + sourceMapContent
}

/**
 * Update the source map contents for improved output.
 */
const updateSourceMap = (sourceMapText: string, normalizedFileName: string): string => {
  const sourceMap = JSON.parse(sourceMapText)
  sourceMap.file = normalizedFileName
  sourceMap.sources = [normalizedFileName]
  delete sourceMap.sourceRoot

  return JSON.stringify(sourceMap)
}

/**
 * Get the file name for the cache entry.
 */
const getCacheName = (sourceCode: string, normalizedFileName: string): string => {
  return sha1(normalizedFileName, '\x00', sourceCode)
}

/**
 * Ensure the given cached content is valid by sniffing for a base64 encoded '}'
 * at the end of the content, which should exist if there is a valid sourceMap present.
 */
const isValidCacheContent = (contents: string): boolean => {
  return /(?:9|0=|Q==)$/.test(contents.slice(-3))
}

/**
 * Compile files which are provided by jest via transform config and cache the result in file system if users run with
 * cache mode
 */
const compileAndCacheResult = (
  cachedir: string | undefined,
  memoryCache: MemoryCache,
  compileFn: CompileFn,
  getExtension: (fileName: string) => string,
  logger: Logger,
) => {
  if (!cachedir) {
    return (code: string, fileName: string, lineOffset?: number) => {
      const normalizedFileName = normalize(fileName)

      logger.debug({ normalizedFileName }, 'readThrough(): no cache')

      const [value, sourceMap] = compileFn(code, normalizedFileName, lineOffset)
      const output = updateOutput(value, fileName, sourceMap, getExtension)
      memoryCache.outputs[normalizedFileName] = output

      return output
    }
  }

  // Make sure the cache directory exists before continuing.
  mkdirp.sync(cachedir)
  try {
    const resolvedModulesCache = readFileSync(getResolvedModulesCache(cachedir), 'utf-8')
    /* istanbul ignore next (covered by e2e) */
    memoryCache.resolvedModules = JSON.parse(resolvedModulesCache)
  } catch (e) {}

  return (code: string, fileName: string, lineOffset?: number) => {
    const normalizedFileName = normalize(fileName)
    const cachePath = join(cachedir, getCacheName(code, normalizedFileName))
    const extension = getExtension(normalizedFileName)
    const outputPath = `${cachePath}${extension}`
    try {
      const output = readFileSync(outputPath, 'utf8')
      if (isValidCacheContent(output)) {
        logger.debug({ normalizedFileName }, 'readThrough(): cache hit')
        memoryCache.outputs[normalizedFileName] = output

        return output
      }
    } catch (err) {}

    logger.debug({ fileName }, 'readThrough(): cache miss')

    const [value, sourceMap] = compileFn(code, normalizedFileName, lineOffset)
    const output = updateOutput(value, normalizedFileName, sourceMap, getExtension)

    logger.debug({ normalizedFileName, outputPath }, 'readThrough(): writing caches')

    memoryCache.outputs[normalizedFileName] = output
    writeFileSync(outputPath, output)

    return output
  }
}

/**
 * Register TypeScript compiler instance.
 * @internal
 */
export const createCompilerInstance = (configs: ConfigSet): TsCompiler => {
  const logger = configs.logger.child({ namespace: 'ts-compiler' })
  const {
    typescript: { options: compilerOptions, fileNames },
    tsJest,
  } = configs
  const cachedir = configs.tsCacheDir
  const ts = configs.compilerModule // Require the TypeScript compiler and configuration.
  const extensions = ['.ts', '.tsx']
  const memoryCache: MemoryCache = {
    contents: Object.create(null),
    versions: Object.create(null),
    outputs: Object.create(null),
    resolvedModules: Object.create(null),
    files: new Map<string, TSFile>(),
  }
  // Enable `allowJs` when flag is set.
  if (compilerOptions.allowJs) {
    extensions.push('.js')
    extensions.push('.jsx')
  }
  // Initialize files from TypeScript into project.
  for (const path of fileNames) {
    const normalizedFilePath = normalize(path)
    memoryCache.versions[normalizedFilePath] = 1
    memoryCache.files.set(normalizedFilePath, {
      version: 0,
    })
  }
  /**
   * Get the extension for a transpiled file.
   */
  const getExtension =
    compilerOptions.jsx === ts.JsxEmit.Preserve
      ? (path: string) => (/\.[tj]sx$/.test(path) ? '.jsx' : '.js')
      : (_: string) => '.js'
  let compilerInstance: CompilerInstance
  if (!tsJest.isolatedModules) {
    // Use language services by default
    compilerInstance = !tsJest.compilerHost
      ? initializeLanguageServiceInstance(configs, memoryCache, logger)
      : initializeProgramInstance(configs, memoryCache, logger)
  } else {
    compilerInstance = initializeTranspilerInstance(configs, memoryCache, logger)
  }
  const compile = compileAndCacheResult(cachedir, memoryCache, compilerInstance.compileFn, getExtension, logger)

  return { cwd: configs.cwd, compile, program: compilerInstance.program }
}
