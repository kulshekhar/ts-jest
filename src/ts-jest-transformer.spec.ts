import { LogLevels } from 'bs-logger'

import { TsJestTransformer } from './ts-jest-transformer'
import { ConfigSet } from './config/config-set'
import { SOURCE_MAPPING_PREFIX } from './compiler/instance'
import { logTargetMock } from './__helpers__/mocks'

const logTarget = logTargetMock()

beforeEach(() => {
  logTarget.clear()
})

describe('TsJestTransformer', () => {
  describe('createOrResolveTransformerCfg', () => {
    it('should return the same config-set for same values with jest config string is not in cachedConfigSets', () => {
      const obj1 = { cwd: '/foo/.' } as any
      const tjT = new TsJestTransformer()

      logTarget.clear()

      tjT.getCacheKey('foo', 'foo', JSON.stringify(obj1), {
        config: obj1,
      } as any)

      expect(logTarget.lines[0]).toMatchInlineSnapshot(`
        "[level:30] no matching config-set found, creating a new one
        "
      `)
    })

    it('should return the same config-set for same values with jest config string in cachedConfigSets', () => {
      const obj1 = { cwd: '/foo/.', rootDir: '/bar//dummy/..', globals: {} }
      const obj2 = { ...obj1 }
      const tjT1 = new TsJestTransformer()
      const tjT2 = new TsJestTransformer()

      logTarget.clear()

      tjT1.getCacheKey('foo', 'foo', JSON.stringify(obj1), {
        config: obj1,
      } as any)
      tjT2.getCacheKey('foo', 'foo', JSON.stringify(obj2), {
        config: obj2,
      } as any)

      expect(logTarget.filteredLines(LogLevels.info)).toHaveLength(1)
    })
  })

  describe('getCacheKey', () => {
    it('should be different for each argument value', () => {
      const tr = new TsJestTransformer()
      const input = {
        fileContent: 'export default "foo"',
        fileName: 'foo.ts',
        jestConfigStr: '{"foo": "bar"}',
        options: { config: { foo: 'bar' } as any, instrument: false, rootDir: '/foo' },
      }
      const keys = [
        tr.getCacheKey(input.fileContent, input.fileName, input.jestConfigStr, input.options),
        tr.getCacheKey(input.fileContent, 'bar.ts', input.jestConfigStr, input.options),
        tr.getCacheKey(input.fileContent, input.fileName, '{}', { ...input.options, instrument: true }),
        tr.getCacheKey(input.fileContent, input.fileName, '{}', { ...input.options, rootDir: '/bar' }),
      ]
      // each key should have correct length
      for (const key of keys) {
        expect(key).toHaveLength(32)
      }
    })
  })

  describe('process', () => {
    let tr: TsJestTransformer

    beforeEach(() => {
      tr = new TsJestTransformer()
    })

    it('should process input as stringified content with content matching stringifyContentPathRegex option', () => {
      const fileContent = '<h1>Hello World</h1>'
      const filePath = 'foo.html'
      const jestCfg = {
        globals: {
          'ts-jest': {
            stringifyContentPathRegex: '\\.html$',
          },
        },
      } as any
      tr.getCacheKey(fileContent, filePath, JSON.stringify(jestCfg), { config: jestCfg } as any)

      const result = tr.process(fileContent, filePath, jestCfg)

      expect(result).toMatchInlineSnapshot(`"module.exports=\\"<h1>Hello World</h1>\\""`)
    })

    it('should process type definition input', () => {
      const fileContent = 'type Foo = number'
      const filePath = 'foo.d.ts'
      const jestCfg = Object.create(null)
      tr.getCacheKey(fileContent, filePath, JSON.stringify(jestCfg), { config: jestCfg } as any)
      const result = tr.process(fileContent, filePath, jestCfg)

      expect(result).toEqual('')
    })

    it('should process js file with allowJs false and show warning log', () => {
      const fileContent = 'const foo = 1'
      const filePath = 'foo.js'
      const jestCfg = {
        globals: {
          'ts-jest': { tsconfig: { allowJs: false } },
        },
      } as any
      tr.getCacheKey(fileContent, filePath, JSON.stringify(jestCfg), { config: jestCfg } as any)
      logTarget.clear()

      const result = tr.process(fileContent, filePath, jestCfg)

      expect(result).toEqual(fileContent)
      expect(logTarget.lines[1].substring(0)).toMatchInlineSnapshot(`
        "[level:40] Got a \`.js\` file to compile while \`allowJs\` option is not set to \`true\` (file: foo.js). To fix this:
          - if you want TypeScript to process JS files, set \`allowJs\` to \`true\` in your TypeScript config (usually tsconfig.json)
          - if you do not want TypeScript to process your \`.js\` files, in your Jest config change the \`transform\` key which value is \`ts-jest\` so that it does not match \`.js\` files anymore
        "
      `)
    })

    it.each(['foo.ts', 'foo.tsx'])('should process ts/tsx file', (filePath) => {
      const fileContent = 'const foo = 1'
      const output = 'var foo = 1'
      const jestCfg = Object.create(null)
      tr.getCacheKey(fileContent, filePath, JSON.stringify(jestCfg), { config: jestCfg } as any)
      jest.spyOn(ConfigSet.prototype, 'tsCompiler', 'get').mockImplementationOnce(() => ({
        compile: () => output,
        cwd: '.',
        program: undefined,
      }))

      const result = tr.process(fileContent, filePath, jestCfg)

      expect(result).toEqual(output)
    })

    it.each(['foo.js', 'foo.jsx'])('should process js/jsx file with allowJs true', (filePath) => {
      const fileContent = 'const foo = 1'
      const output = 'var foo = 1'
      const jestCfg = {
        globals: {
          'ts-jest': { tsconfig: { allowJs: true } },
        },
      } as any
      tr.getCacheKey(fileContent, filePath, JSON.stringify(jestCfg), { config: jestCfg } as any)
      logTarget.clear()
      jest.spyOn(ConfigSet.prototype, 'tsCompiler', 'get').mockImplementationOnce(() => ({
        compile: () => output,
        cwd: '.',
        program: undefined,
      }))

      const result = tr.process(fileContent, filePath, jestCfg)

      expect(result).toEqual(output)
    })

    it('should process file with unknown extension and show warning message without babel-jest', () => {
      const fileContent = 'foo'
      const filePath = 'foo.bar'
      const jestCfg = {
        globals: {
          'ts-jest': { tsconfig: { allowJs: true } },
        },
      } as any
      tr.getCacheKey(fileContent, filePath, JSON.stringify(jestCfg), { config: jestCfg } as any)
      logTarget.clear()

      const result = tr.process(fileContent, filePath, jestCfg)

      expect(result).toEqual(fileContent)
      expect(logTarget.lines[1]).toMatchInlineSnapshot(`
        "[level:40] Got a unknown file type to compile (file: foo.bar). To fix this, in your Jest config change the \`transform\` key which value is \`ts-jest\` so that it does not match this kind of files anymore.
        "
      `)
    })

    it.each(['foo.bar', 'foo.js'])('should process file with babel-jest', (filePath) => {
      const fileContent = 'foo'
      const jestCfg = {
        globals: {
          'ts-jest': { babelConfig: true },
        },
      } as any
      tr.getCacheKey(fileContent, filePath, JSON.stringify(jestCfg), { config: jestCfg } as any)
      logTarget.clear()

      const result = tr.process('foo', filePath, jestCfg)

      if (typeof result !== 'string') {
        expect(result.code.substring(0, result.code.indexOf(SOURCE_MAPPING_PREFIX))).toMatchSnapshot()
      }
      if (filePath === 'foo.bar') {
        expect(logTarget.filteredLines(LogLevels.warn)[0]).toMatchSnapshot()
      }
    })
  })
})
