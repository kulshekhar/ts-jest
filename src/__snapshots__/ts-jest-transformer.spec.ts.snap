// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TsJestTransformer process should process file with babel-jest 1`] = `
"foo;
//# "
`;

exports[`TsJestTransformer process should process file with babel-jest 2`] = `
"[level:40] Got a unknown file type to compile (file: foo.bar). To fix this, in your Jest config change the \`transform\` key which value is \`ts-jest\` so that it does not match this kind of files anymore. If you still want Babel to process it, add another entry to the \`transform\` option with value \`babel-jest\` which key matches this type of files.
"
`;

exports[`TsJestTransformer process should process file with babel-jest 3`] = `
"foo;
//# "
`;
